File: app/backend/services/gastronom_loyalty.py
"""Loyalty gifts for ГАСТРАНОМ — free gifts by order subtotal tiers."""

from __future__ import annotations

import json
import uuid
from typing import Any, Dict, List, Optional


def _parse_float(value: Any, default: float = 0.0) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return default


def is_loyalty_enabled(settings: Dict[str, str]) -> bool:
    raw = (settings.get("loyalty_enabled") or "1").strip().lower()
    return raw not in ("0", "false", "no", "off")


def parse_loyalty_gifts(settings: Dict[str, str]) -> List[Dict[str, Any]]:
    raw = settings.get("loyalty_gifts") or "[]"
    try:
        data = json.loads(raw) if isinstance(raw, str) else raw
    except json.JSONDecodeError:
        return []
    if not isinstance(data, list):
        return []

    gifts: List[Dict[str, Any]] = []
    for idx, item in enumerate(data):
        if not isinstance(item, dict):
            continue
        title = str(item.get("title") or "").strip()
        if not title:
            continue
        min_amount = _parse_float(item.get("min_amount"), 0)
        if min_amount <= 0:
            continue
        gifts.append({
            "id": str(item.get("id") or f"gift-{idx + 1}"),
            "min_amount": min_amount,
            "title": title,
            "description": str(item.get("description") or "").strip(),
            "image_url": str(item.get("image_url") or "").strip(),
            "is_active": item.get("is_active") is not False,
            "sort_order": int(item.get("sort_order") or idx + 1),
        })

    gifts.sort(key=lambda g: (g["min_amount"], g["sort_order"]))
    return gifts


def resolve_loyalty_gift(
    subtotal: float,
    settings: Dict[str, str],
) -> Optional[Dict[str, Any]]:
    """Best active gift tier reached by cart subtotal (products only, no delivery)."""
    if not is_loyalty_enabled(settings):
        return None
    gifts = [g for g in parse_loyalty_gifts(settings) if g.get("is_active")]
    if not gifts:
        return None
    matched: Optional[Dict[str, Any]] = None
    for gift in gifts:
        if subtotal >= gift["min_amount"]:
            matched = gift
    return matched


def next_loyalty_gift(
    subtotal: float,
    settings: Dict[str, str],
) -> Optional[Dict[str, Any]]:
    if not is_loyalty_enabled(settings):
        return None
    gifts = [g for g in parse_loyalty_gifts(settings) if g.get("is_active")]
    for gift in gifts:
        if subtotal < gift["min_amount"]:
            return gift
    return None


def gift_order_line(gift: Dict[str, Any]) -> Dict[str, Any]:
    min_amount = int(gift["min_amount"])
    title = gift["title"]
    return {
        "id": None,
        "name": f"🎁 Подарок: {title}",
        "weight": "",
        "qty": 1,
        "price": 0,
        "sum": 0,
        "is_gift": True,
        "gift_id": gift.get("id"),
        "gift_min_amount": min_amount,
    }


def gift_comment_line(gift: Dict[str, Any]) -> str:
    min_amount = int(gift["min_amount"])
    return f"Подарок: {gift['title']} (от {min_amount:,} ₸)".replace(",", " ")


def default_loyalty_gifts_json() -> str:
    gifts = [
        {
            "id": str(uuid.uuid4()),
            "min_amount": 5000,
            "title": "Ручка с логотипом",
            "description": "Фирменная ручка в подарок",
            "image_url": "",
            "is_active": True,
            "sort_order": 1,
        },
        {
            "id": str(uuid.uuid4()),
            "min_amount": 10000,
            "title": "Чупа-чупс",
            "description": "Сладкий подарок к заказу",
            "image_url": "",
            "is_active": True,
            "sort_order": 2,
        },
        {
            "id": str(uuid.uuid4()),
            "min_amount": 15000,
            "title": "Набор сладостей",
            "description": "Несколько сладостей в подарок",
            "image_url": "",
            "is_active": True,
            "sort_order": 3,
        },
    ]
    return json.dumps(gifts, ensure_ascii=False)

File: app/frontend/src/lib/gastronomLoyalty.ts
export interface LoyaltyGift {
  id: string;
  min_amount: number;
  title: string;
  description?: string;
  image_url?: string;
  is_active: boolean;
  sort_order: number;
}

export function isLoyaltyEnabled(settings: Record<string, string | undefined>): boolean {
  const raw = (settings.loyalty_enabled ?? '1').trim().toLowerCase();
  return !['0', 'false', 'no', 'off'].includes(raw);
}

export function parseLoyaltyGifts(raw: string | undefined): LoyaltyGift[] {
  if (!raw) return [];
  try {
    const data = JSON.parse(raw);
    if (!Array.isArray(data)) return [];
    return data
      .map((item, idx) => {
        if (!item || typeof item !== 'object') return null;
        const title = String(item.title || '').trim();
        const minAmount = Number(item.min_amount) || 0;
        if (!title || minAmount <= 0) return null;
        return {
          id: String(item.id || `gift-${idx + 1}`),
          min_amount: minAmount,
          title,
          description: String(item.description || '').trim(),
          image_url: String(item.image_url || '').trim(),
          is_active: item.is_active !== false,
          sort_order: Number(item.sort_order) || idx + 1,
        } satisfies LoyaltyGift;
      })
      .filter(Boolean)
      .sort((a, b) => a!.min_amount - b!.min_amount || a!.sort_order - b!.sort_order) as LoyaltyGift[];
  } catch {
    return [];
  }
}

export function serializeLoyaltyGifts(gifts: LoyaltyGift[]): string {
  return JSON.stringify(
    gifts.map((g, idx) => ({
      ...g,
      sort_order: g.sort_order || idx + 1,
    }))
  );
}

export function resolveLoyaltyGift(subtotal: number, gifts: LoyaltyGift[]): LoyaltyGift | null {
  const active = gifts.filter((g) => g.is_active);
  let matched: LoyaltyGift | null = null;
  for (const gift of active) {
    if (subtotal >= gift.min_amount) matched = gift;
  }
  return matched;
}

export function nextLoyaltyGift(subtotal: number, gifts: LoyaltyGift[]): LoyaltyGift | null {
  const active = gifts.filter((g) => g.is_active);
  return active.find((g) => subtotal < g.min_amount) ?? null;
}

export function newLoyaltyGift(index: number): LoyaltyGift {
  return {
    id: crypto.randomUUID?.() || `gift-${Date.now()}`,
    min_amount: (index + 1) * 5000,
    title: '',
    description: '',
    image_url: '',
    is_active: true,
    sort_order: index + 1,
  };
}

export function formatMoney(n: number) {
  return `${Math.round(n).toLocaleString('ru-RU')} ₸`;
}
